iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/newtype/decrementable.hpp

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_CXX_NEWTYPE_DECREMENTABLE_HPP
+#define IOX_HOOFS_CXX_NEWTYPE_DECREMENTABLE_HPP
+#include "iceoryx_hoofs/internal/cxx/newtype/internal.hpp"
+
+namespace iox
+{
+namespace cxx
+{
+namespace newtype
+{
+template <typename T>
+// not required since a default'ed destructor does not define a destructor, hence the copy/move operations are
+// not deleted.
+// the only adaptation is that the dtor is protected to prohibit the user deleting the child type by
+// explicitly calling the destructor of the base type.
+// NOLINTNEXTLINE(cppcoreguidelines-special-member-functions, hicpp-special-member-functions)
+struct Decrementable
+{
+  protected:
+    ~Decrementable() = default;
+};
+
+template <typename T>
+auto operator--(T& rhs) noexcept -> typename T::value_type
+{
+    return internal::preDecrement(rhs);
+}
+
+template <typename T>
+auto operator--(T& rhs, int) noexcept -> typename T::value_type
+{
+    auto value = internal::newTypeAccessor(rhs);
+    internal::preDecrement(rhs);
+    return value;
+}
```

Comment:
t.b.h. I would not add this trait. It is just a convenient function for `+=1` which does lead to problems like `foo(++i, ++i)` not doing what one would expect. That said, I'd rather see `operator+=`, `operator-=`, etc. which also gives one the benefit to increment by more than 1.

Reply:
@elBoberido You have a point there. However, when adding `NewType` to legacy code, I found this trait quite useful to make loops compile without any code changes.

Reply:
@elBoberido Here I have to say that I disagree and side with @mossmaurice .

When one writes `foo(++i, ++i);` (which in itself is weird and ugly) and no one can assume that the first argument is smaller than the second (there is no guarantee since the argument evaluation order is unspecified, see: https://en.cppreference.com/w/cpp/language/eval_order)
or that `i` is unchanged.

Therefore, I would state that with this operator C++ does not behave any weirder or more error prone than before :/ and from my view it is alright to add this.
I agree with the problem you see here but for me it is a problem of the language which we cannot solve.

Reply:
@elfenpiff well, exceptions are also part of the language but we decided not to use them. I still think this gives little value for the problems it can create.